Integration tests for KubernetesReservationClient; trying to reproduce different LOST cases

dsvoronin opened the pull request:
there is still not much progress, but it's complete chunk of work, ready for merge

dsvoronin committed:
cosmetics

dsvoronin committed:
integration test for KubernetesReservationClient

dsvoronin commented on subprojects/gradle/instrumentation-changed-tests-finder/src/main/kotlin/com/avito/android/FindChangedTestsTask.kt:
```diff
@@ -19,7 +19,7 @@ import javax.inject.Inject
  */
 @Suppress("UnstableApiUsage")
-open class FindChangedTestsTask @Inject constructor(
+abstract class FindChangedTestsTask @Inject constructor(
```
leftover for #669

dsvoronin committed:
throw exception on incorrect image reference

dsvoronin committed:
Merge branch 'develop' into MBS-8662-lost

dsvoronin committed:
unknown host case test

dsvoronin committed:
remove verbose logs

dsvoronin committed:
test for deployment duplicates

dsvoronin committed:
fix detekt

dsvoronin committed:
pass parameters

dsvoronin committed:
Merge branch 'develop' into MBS-8662-lost

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/Utils.kt:
```diff
@@ -0,0 +1,3 @@
+package com.avito.instrumentation.reservation.client.kubernetes
+
+fun String.kubernetesName(): String = replace("_", "-").toLowerCase()
```
Can it be at least internal?

dsvoronin replied:
yes, thanks;
really struggling without lib mode; hope gradle 6.8 will be enough to force 1.4 kotlin in project

dsvoronin replied:
Moved to `:kubernetes` module, which is implementation dependency for `instrumentation-tests`

can't use `internal` because it is accessed from various places. like test fixtures

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt:
```diff
@@ -269,6 +273,11 @@ class KubernetesReservationClient(
             // https://kubernetes.io/docs/tasks/debug-application-cluster/debug-application/#my-pod-stays-waiting
             if (!waitingMessage.isNullOrBlank()) {
                 logger.warn("Can't start pod on this node: $waitingMessage")
```
What is "node" in the message? Is it k8s node?

dsvoronin replied:
yes, but i don't understand why it's important here (probably same as you)

eugene-krivobokov replied:
Ok, let's rephrase to `Can't start pod: ...`

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/executing/TestExecutor.kt:
```diff
@@ -37,8 +38,7 @@ interface TestExecutor {
         private val runner = TestsRunnerClient()
 
-        // todo hi Juno!
-        private val outputDirectoryName = "composer"
```
:tada:

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClientIntegrationTest.kt:
```diff
@@ -0,0 +1,173 @@
+import org.junit.jupiter.api.assertThrows
+import java.io.File
+import java.net.UnknownHostException
+
```
What about [`KubernetesServer`](https://github.com/fabric8io/kubernetes-client#mocking-kubernetes) mocks?
Will it be helpful?

dsvoronin replied:
If i understand correctly, you're proposing to record k8s responses and convert this tests to hermetic?
i am hands down like this approach, but want to keep flag to switch between integration and mock/stub variant;
I see a value here in testing migration on newer k8s versions, that could behave differently

eugene-krivobokov replied:
I didn't mean a total replacement in this PR, just wondering about plans on this topic.

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClientIntegrationTest.kt:
```diff
@@ -0,0 +1,173 @@
+        kubernetesNamespace: String = requireNotNull(System.getProperty("avito.kubernetes.namespace")),
+        configurationName: String = "integration-test",
+        projectName: String = "",
+        buildId: String = "19723577",
```
How about "-1" to make a more test-ish value?

dsvoronin replied:
will do, agreed

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClientIntegrationTest.kt:
```diff
@@ -0,0 +1,173 @@
+import java.io.File
+import java.net.UnknownHostException
+
+internal class KubernetesReservationClientIntegrationTest {
```
What if we run multiple integration tests in parallel in different PRs? Is it ok?

dsvoronin replied:
good catch, could be problems without some uniquiness in deployment name; will try to add some PR info

dsvoronin replied:
Fixed it via buildId parameter from teamcity run in deploymentName

sergioboi commented on subprojects/detekt.yml:
```diff
@@ -620,7 +620,7 @@ style:
         ignoreHashCodeFunction: true
         ignorePropertyDeclaration: true
-        ignoreLocalVariableDeclaration: false
+        ignoreLocalVariableDeclaration: true
```
What does it affect?

dsvoronin replied:
extracting magic value to local variable didn't help with this flag disabled.

```
fun someFun() {
   someExpression(42) < -- MAGIC NUMBER
}
```

fix that didn't work before setting `ignoreLocalVariableDeclaration` to `true`:

```
fun someFun() {
   val meaningfulValueName = 42 < -- WAS MAGIC NUMBER STILL
   someExpression(meaningfulValueName) 
}
```

sergioboi commented on subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/FakeDeploymentNameGenerator.kt:
```diff
@@ -0,0 +1,9 @@
+package com.avito.instrumentation.reservation.client.kubernetes
+
+class FakeDeploymentNameGenerator(
+    private val variablePart: String = "integration-test"
```
I don't understand the name `variablePart`

dsvoronin replied:
let's rename it then;
how about `deploymentName`, `name`, `customName`?

sergioboi replied:
`deploymentNamePrefix`?

dsvoronin replied:
`deploymentName` already in class name, `prefix` is in fact `postfix`

sergioboi commented on subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClientIntegrationTest.kt:
```diff
@@ -0,0 +1,173 @@
+
+        val exception = assertThrows<KubernetesClientException> {
+            runBlocking {
+                clientOne!!.claim(
```
Why do we need to force unwrap?

dsvoronin replied:
because it's nullable type; couldn't use lateinit here

sergioboi commented on subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClientIntegrationTest.kt:
```diff
@@ -0,0 +1,173 @@
+        val delayEnoughToCreateFirstDeploymentMs = 3000L
+
+        val exception = assertThrows<KubernetesClientException> {
+            runParallel(
```
Could we write a test without parallelism and delay?

dsvoronin replied:
i don't know how to test such thing in integration, and i explained why i want to keep integration variant of tests here: https://github.com/avito-tech/avito-android/pull/673#discussion_r542305675

sergioboi replied:
I agree that we need a real interaction, but I don't know do we need it parallel?
If we use the synchronous dispatcher will that help you?

dsvoronin replied:
i need two different clients (look at them as different PR jobs) to create deployment with same name one after another;
i'm not sure if sync dispatcher will help me;
if you know how to do it - please help

sergioboi commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt:
```diff
@@ -269,6 +273,11 @@ class KubernetesReservationClient(
+
+                // handle special cases
+                if (waitingMessage.contains("couldn't parse image reference")) {
+                    error("Can't create pods for deployment, check image reference: $waitingMessage")
```
Do we have any reference to cause exception?

dsvoronin replied:
what do you mean with `reference` here?

sergioboi replied:
On reference, I mean java object `Throwable`

dsvoronin committed:
move and rename toValidKubernetesName

dsvoronin committed:
Merge branch 'develop' into MBS-8662-lost

dsvoronin committed:
fix log message

dsvoronin committed:
move test factory KubernetesReservationClient to test fixtures module

dsvoronin committed:
rename FakeDeploymentNameGenerator param

dsvoronin committed:
fix concurrent PR's integration check problem

introduce TestProperty to deal with empty and null values passed from Gradle via systemProperties

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/integTest/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClientIntegrationTest.kt:
```diff
@@ -0,0 +1,132 @@
+import org.junit.jupiter.api.assertThrows
+import java.net.UnknownHostException
+
+internal class KubernetesReservationClientIntegrationTest {
```
Should "Clear leaked K8S deployments" know about these fake deployments?

dsvoronin replied:
yes, in theory this logic about deployments should be shared;
i see "clear leaked deployments" as obsolete thing after introducing shared service, but for now it must be done
thanks

dsvoronin committed:
Merge branch 'develop' into MBS-8662-lost

dsvoronin committed:
support ClearK8SDeployments for integration tests

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests-test-fixtures/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/FakeDeploymentNameGenerator.kt:
```diff
@@ -0,0 +1,11 @@
+import com.avito.utils.gradle.toValidKubernetesName
+
+class FakeDeploymentNameGenerator(
+    private val buildId: String
```
Do we need to know a real `buildId`? 
It seems the random value is enough.

dsvoronin replied:
yep, fixed

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/KubernetesReservationClient.kt:
```diff
@@ -320,6 +330,8 @@ class KubernetesReservationClient(
         object Idling : State()
     }
+
+    companion object
```
What is this?

dsvoronin replied:
my "signature" companion object for `createStubInstance()` function in test fixtures module

eugene-krivobokov commented on subprojects/gradle/kubernetes/src/main/kotlin/com/avito/utils/gradle/Naming.kt:
```diff
@@ -0,0 +1,8 @@
+package com.avito.utils.gradle
+
+/**
+ * if you encounter any problems with generated named, feel free to add more strict checks and changes
```
This is true for any code

dsvoronin replied:
fair enough; deleted

eugene-krivobokov commented on subprojects/gradle/test-project/src/main/kotlin/com/avito/test/gradle/TestProperty.kt:
```diff
@@ -0,0 +1,19 @@
+ * property passed through gradle (Test task).systemProperty
+ * see build.gradle.kts
+ */
+fun getTestProperty(name: String, defaultValue: String? = null): String {
```
```suggestion
fun getSystemProperty(name: String, defaultValue: String? = null): String {
```
I see nothing about tests in implementation

dsvoronin replied:
fixed; also moved to kotlin-dsl-support

eugene-krivobokov commented on subprojects/gradle/instrumentation-tests/src/main/kotlin/com/avito/instrumentation/reservation/client/kubernetes/ReservationDeploymentFactory.kt:
```diff
@@ -214,17 +220,17 @@ class ReservationDeploymentFactory(
+             * for incorrectly finished builds
+             * see [com.avito.ci.DeploymentEnvironment]
+             */
+            "type" to buildType,
```
It'd good to get this value from the code. Otherwise, it's easy to break this contract in `DeploymentEnvironment`

dsvoronin replied:
that's right, but this modules not linked; and introducting new "bridge" module only for pair of constants seems overkill for me;

long term support will not be an issue because of new service for deployments management

dsvoronin committed:
random postfix in deploymentName generator

but ability to fixate it per generator instance for tests

dsvoronin committed:
remove useless comment

dsvoronin committed:
rename getTestProperty -> getSystemProperty

moved to kotlin-dsl-support
